cmd/github.go

```diff
@@ -0,0 +1,351 @@
+type githubAuthorizationRequest struct {
+	Scopes      []string `json:"scopes"`
+	Notes       string   `json:"note"`
+	Fingerprint string   `json:"fingerprint"`
+}
+
+// githubAuthorizationResponse is the response that is given for Personal Authorization Token
+// requests
+type githubAuthorizationResponse struct {
+	ID    int
+	Token string
+
+	Errors []struct {
+		Code string
+	}
+}
+
+// githubErrors are the human readable error messages from GitHub during the Personal Authorization
+// Token flow
+var githubErrors = map[string]string{
+	"already_exists": "A Personal Authorization Token already exists for Manifold",
+}
+
+// oauthStoreFunc is responsible for storing the OAuth authentication with Manifold, either by
+// linking accounts or logging in
+type oauthStoreFunc func(ctx context.Context, cfg *config.Config, a *analytics.Analytics, req models.OAuthAuthenticationRequest) error
+
+// githubWithCallback identified with GitHub via. the normal browser-based OAuth flow
+func githubWithCallback(ctx context.Context, cfg *config.Config, a *analytics.Analytics, store oauthStoreFunc) error {
+	authConfig := &oauth2.Config{
+		ClientID:    config.GitHubClientID,
+		Scopes:      []string{"user"},
+		Endpoint:    github.Endpoint,
+		RedirectURL: "http://127.0.0.1:49152/github/callback",
+	}
+
+	// set up the oauth client
+	state := genRandomString(12)
+	url := authConfig.AuthCodeURL(state, oauth2.AccessTypeOffline)
+
+	// let the user auth in the browser
+	authMsg := color.Color(ansiterm.Blue,
+		"A browser window will open for authentication. Please close this browser when finished")
+	fmt.Println(authMsg)
+
+	time.Sleep(time.Second * 2) // give them time to read the message
+
+	err := open.Start(url)
+	if err != nil {
+		return cli.NewExitError("Unable to open authorization in browser", -1)
+	}
+
+	// handle http callback
+	done := make(chan bool)
+	errCh := make(chan error)
+	go func() {
+		http.HandleFunc("/github/callback", githubCallback(ctx, cfg, state, store, a, done, errCh))
+
+		errCh <- http.ListenAndServe(":49152", nil)
+	}()
+
+	select {
+	case <-done:
+		authedMsg := color.Color(ansiterm.Green, "You are now authenticated with GitHub, and can close your browser window")
+		fmt.Printf("\n%s\n", authedMsg)
+	case err := <-errCh:
+		errMsg := color.Color(ansiterm.Red, fmt.Sprintf("Error with authentication: %s", err))
+		return errors.New(errMsg)
+	}
+
+	return nil
+}
+
+// githubCallback is the OAuth callback function for the browser-based flow
+func githubCallback(ctx context.Context, cfg *config.Config, state string, store oauthStoreFunc,
+	a *analytics.Analytics, done chan bool, errCh chan error) func(w http.ResponseWriter, r *http.Request) {
+
+	badRequest := func(w http.ResponseWriter, msg string, vals ...interface{}) {
+		valMsg := fmt.Sprintf(msg, vals...)
+		fmtMsg := fmt.Sprintf("<html><body><h2>Bad request</h2><p>%s</p></body>", valMsg)
```

Comment:
This is probably fine to start, but we should open an issue and add something to the Design backlog to bring some basic CSS to this flow.

Reply:
Added this ticket for discussion: https://github.com/manifoldco/engineering/issues/3404